File management Service

MostafaFouad0 opened the pull request:
**CU-8696636j1**
### Overview
This PR introduces a microservice that allows users to upload files. The service stores these files on disk and responds with a unique file ID for retrieval.

### Features Implemented 
- File Upload Endpoint: Accepts an image file via a POST request, stores it on the server and store it's path in a database.
- File Retrieval Endpoint: Gives a way to retrieve an image given its ID.
### Future Work
- Check if a file is malicious
- I will try to use caching to enhance the performance

MostafaFouad0 committed:
implemented file table in database using Prisma client

MostafaFouad0 committed:
implemented the endpoint for receiving and locally storing files

MostafaFouad0 committed:
updated .gitignore

MostafaFouad0 committed:
implemented endpoint to send a file given it's ID.

MostafaFouad0 committed:
added documentation in README file

Mohab96 commented:
Task linked: [CU-8696636j1 Files Management Component](https://app.clickup.com/t/8696636j1)

Mohab96 commented on src/services/File Management/controllers/filesControllers.js:
```diff
@@ -0,0 +1,58 @@
+const storeFile = async (req, res) => {
+  try {
+    if (!req.file) {
+      return res.status(400).json({
```
For the sake of consistency, try wrapping any response with wrapper like [this one](https://github.com/Mohab96/online-chess-game/blob/master/src/utils/apiResponse.js) so that when somebody builds a client it can check the status of the response before going into the actual data of the response.

Take [this](https://www.npmjs.com/package/jsend) as a reference.

Mohab96 commented on src/services/File Management/controllers/filesControllers.js:
```diff
@@ -0,0 +1,58 @@
+    });
+  } catch (error) {
+    console.error("Error uploading files:", error);
+    return res.status(500).json({
```
Try using status codes like [this one](https://github.com/Mohab96/online-chess-game/blob/master/src/utils/statusCodes.js) as it's better than writing the status codes explicitly into the code like this.

Mohab96 commented on src/services/File Management/controllers/filesControllers.js:
```diff
@@ -0,0 +1,58 @@
+const sendFile = async (req, res) => {
+  try {
+    const fileId = +req.params.fileId;
+    const file = await prisma.file.findFirst({
```
Please validate if `fileId` is null or not before querying the DB.

Mohab96 commented on src/services/File Management/server.js:
```diff
@@ -0,0 +1,19 @@
+app.use("/api/file", fileRoutes);
+
+app.listen(process.env.PORT, () => {
+  console.log("[!] server is running on port:", process.env.PORT);
```
replace `process.env.PORT` with `process.env.PORT || 3000` in case the `PORT` wasn't supplied in the `.env` file.

Mohab96 reviewed (changes requested):
Well done Mostafa, please update the PR according to the comments regarding the PORT env variable and the validation of `fileId`. Any other comments can be treated as future work.

MostafaFouad0 committed:
update the schema of the file table

added a new column called status which will store the status of the file e.g. [In Queue, In Progress, Failed], this will be helpfull for to know the status of the file before requesting it or even when trying to retrieve it.

MostafaFouad0 committed:
created rmqConsumerServer.js file

implemented the logic of the consumer server with rabbitMQ

MostafaFouad0 committed:
created statusCodes.js

MostafaFouad0 committed:
added more status codes

MostafaFouad0 committed:
fixed bugs

fixed bug when the process.env.PORT is null

MostafaFouad0 committed:
added more logic when writing a file fail

added a small block of code to update the status of a file to "Failed"

MostafaFouad0 committed:
integrated RabbitMQ

MostafaFouad0 committed:
updated the file size limit

MostafaFouad0 committed:
updated README.md

MostafaFouad0 committed:
updated README.md

MostafaFouad0 commented:
@Mohab96 changes made please take a look

Mohab96 commented:
Great work! Please move the `src/services/File Management/utils/` to `src/shared/utils/` and you're good to go.

MostafaFouad0 committed:
moved utils folder to /src/shared
